File: daySeven/JavaScriptMenstruationCycle/MenstruationCycle.js
// MenstruationCycle.js

const { parseISO, format, addDays, subDays, differenceInDays, isBefore } = require('date-fns');
const prompt = require('prompt-sync')({ sigint: true });

class MenstruationCycle {
    static cycleDetails = [];

    constructor(name, previousPeriodDate, currentPeriodDate, periodDurationDays) {
        this.name = name;
        this.previousPeriodDate = previousPeriodDate;
        this.currentPeriodDate = currentPeriodDate;
        this.periodDurationDays = periodDurationDays;
    }

    getName() {
        return this.name;
    }

    showMenstrualCycle() {
        const periodCycleLength = differenceInDays(this.currentPeriodDate, this.previousPeriodDate);
        const ovulationDate = addDays(this.currentPeriodDate, periodCycleLength - 14);
        const fertileWindowStart = subDays(ovulationDate, 5);
        const fertileWindowEnd = addDays(ovulationDate, 1);
        const peakFertilePeriodStart = subDays(ovulationDate, 2);
        const peakFertilePeriodEnd = addDays(ovulationDate, 2);
        const safePeriodStart1 = this.currentPeriodDate;
        const safePeriodEnd1 = subDays(fertileWindowStart, 1);
        const safePeriodStart2 = addDays(fertileWindowEnd, 1);
        const safePeriodEnd2 = addDays(this.currentPeriodDate, periodCycleLength - 1);
        const pmsStart = addDays(this.currentPeriodDate, periodCycleLength - 10);
        const pmsEnd = addDays(this.currentPeriodDate, periodCycleLength - 1);
        const periodEndDate = addDays(this.currentPeriodDate, this.periodDurationDays - 1);

        console.log(`\n=== User Name ===: ${this.name}`);
        console.log("Here's a breakdown of your Menstrual Cycle");
        console.log(`Your Cycle length is: ${periodCycleLength} days`);

        if (periodCycleLength >= 21 && periodCycleLength <= 35) {
            console.log("Your Period Cycle is within the normal range (21-35 days).");
            console.log("Here is your next 5 predicted period dates for the next 5 months:");

            let nextPeriodDate = this.currentPeriodDate;
            for (let i = 1; i <= 5; i++) {
                console.log(`Period ${i}: ${format(nextPeriodDate, 'yyyy-MM-dd')}`);
                nextPeriodDate = addDays(nextPeriodDate, periodCycleLength);
            }
        } else {
            console.log("Your Cycle is outside the normal range.");
        }

        console.log(`Your Ovulation date is: ${format(ovulationDate, 'yyyy-MM-dd')}`);
        console.log(`Your Fertility Window (full): ${format(fertileWindowStart, 'yyyy-MM-dd')} to ${format(fertileWindowEnd, 'yyyy-MM-dd')}`);
        console.log(`Peak Fertility Days (most fertile): ${format(peakFertilePeriodStart, 'yyyy-MM-dd')} to ${format(peakFertilePeriodEnd, 'yyyy-MM-dd')}`);
        console.log(`Your Safe period before the start of your Fertile Window is: ${format(safePeriodStart1, 'yyyy-MM-dd')} to ${format(safePeriodEnd1, 'yyyy-MM-dd')}`);
        console.log(`Your Safe period after the end of your Fertile Window is: ${format(safePeriodStart2, 'yyyy-MM-dd')} to ${format(safePeriodEnd2, 'yyyy-MM-dd')}`);
        console.log(`PMS days are likely to begin between: ${format(pmsStart, 'yyyy-MM-dd')} to ${format(pmsEnd, 'yyyy-MM-dd')}`);
        console.log(`Your period will likely last: ${format(this.currentPeriodDate, 'yyyy-MM-dd')} to ${format(periodEndDate, 'yyyy-MM-dd')}`);
        console.log(`Your Estimated period duration is: ${this.periodDurationDays} days`);
    }

    static addUserDetails() {
        let name;

        while (true) {
            name = prompt("Please Enter your Name: ").trim();

            const duplicate = this.cycleDetails.find(u => u.name.toLowerCase() === name.toLowerCase());
            if (!duplicate) break;
            console.log(`Error: User with the name '${name}' already exists. Please enter a different name.`);
        }

        let previousPeriodDate, currentPeriodDate, periodDurationDays;

        while (true) {
            try {
                previousPeriodDate = parseISO(prompt("Enter your previous period date (YYYY-MM-DD): ").trim());
                if (isNaN(previousPeriodDate)) throw new Error();
                break;
            } catch {
                console.log("Invalid date format. Please use YYYY-MM-DD.");
            }
        }

        while (true) {
            try {
                currentPeriodDate = parseISO(prompt("Enter your current period date (YYYY-MM-DD): ").trim());
                if (isNaN(currentPeriodDate)) throw new Error();
                if (isBefore(currentPeriodDate, previousPeriodDate)) {
                    console.log("Error: Current date must be after the previous date.");
                    continue;
                }
                break;
            } catch {
                console.log("Invalid date format. Please use YYYY-MM-DD.");
            }
        }

        while (true) {
            try {
                periodDurationDays = parseInt(prompt("Enter your typical period duration (in days): ").trim());
                if (isNaN(periodDurationDays)) throw new Error();
                break;
            } catch {
                console.log("Invalid input for period duration. Please enter a valid number.");
            }
        }

        this.cycleDetails.push(new MenstruationCycle(name, previousPeriodDate, currentPeriodDate, periodDurationDays));
        console.log("Your details are currently being analyzed!! Please check on the View Menstruation Summary Section to view the summary of your Cycle.");
    }

    static viewMenstrualSummary() {
        if (this.cycleDetails.length === 0) {
            console.log("There are no Menstrual Records available.");
            return;
        }

        this.cycleDetails.forEach((user, index) => {
            console.log(`\nUser ${index + 1}`);
            user.showMenstrualCycle();
        });
    }

    static updateMenstruationCycle() {
        if (this.cycleDetails.length === 0) {
            console.log("There are no user records to update.");
            return;
        }

        const indexName = prompt("Please Enter the user name to update Menstrual Summary: ").trim();
        const foundUser = this.cycleDetails.find(user => user.getName().toLowerCase() === indexName.toLowerCase());

        if (!foundUser) {
            console.log("User not found");
            return;
        }

        while (true) {
            try {
                foundUser.previousPeriodDate = parseISO(prompt("Enter the new previous period date (YYYY-MM-DD): ").trim());
                if (isNaN(foundUser.previousPeriodDate)) throw new Error();
                break;
            } catch {
                console.log("Invalid date format. Please use YYYY-MM-DD.");
            }
        }

        while (true) {
            try {
                const newCurrentDate = parseISO(prompt("Enter the new current period date (YYYY-MM-DD): ").trim());
                if (isNaN(newCurrentDate)) throw new Error();
                if (isBefore(newCurrentDate, foundUser.previousPeriodDate)) {
                    console.log("Error: Current date must be after the previous date.");
                    continue;
                }
                foundUser.currentPeriodDate = newCurrentDate;
                break;
            } catch {
                console.log("Invalid date format. Please use YYYY-MM-DD.");
            }
        }

        while (true) {
            try {
                foundUser.periodDurationDays = parseInt(prompt("Enter the new typical period duration (in days): ").trim());
                if (isNaN(foundUser.periodDurationDays)) throw new Error();
                break;
            } catch {
                console.log("Invalid input for period duration. Please enter a valid number.");
            }
        }

        console.log("The User Menstrual Summary has been updated!!");
    }
}

module.exports = MenstruationCycle;

File: daySeven/menstrualWahala/JavaMenstrualCycle/MenstruationCycle.java
import java.util.ArrayList;
import java.util.Scanner;
import java.time.LocalDate;
import java.time.format.DateTimeParseException;
import java.time.temporal.ChronoUnit;

public class MenstruationCycle {
    static Scanner input = new Scanner(System.in);
    static ArrayList<MenstruationCycle> cycleDetails = new ArrayList<>();

    String name;
    LocalDate previousPeriodDate;
    LocalDate currentPeriodDate;
    int periodDurationDays;

    public MenstruationCycle(String name, LocalDate previousPeriodDate, LocalDate currentPeriodDate, int periodDurationDays) {
        this.name = name;
        this.previousPeriodDate = previousPeriodDate;
        this.currentPeriodDate = currentPeriodDate;
        this.periodDurationDays = periodDurationDays;
    }

    public String getName() {
        return name;
    }
    
    public void showMenstrualCycle() {
        long periodCycleLength = ChronoUnit.DAYS.between(previousPeriodDate, currentPeriodDate);

        LocalDate ovulationDate = currentPeriodDate.plusDays(periodCycleLength - 14);
        LocalDate fertileWindowStart = ovulationDate.minusDays(5);
        LocalDate fertileWindowEnd = ovulationDate.plusDays(1);

        LocalDate peakFertilePeriodStart = ovulationDate.minusDays(2);
        LocalDate peakFertilePeriodEnd = ovulationDate.plusDays(2);

        LocalDate safePeriodStart1 = currentPeriodDate;
        LocalDate safePeriodEnd1 = fertileWindowStart.minusDays(1);

        LocalDate safePeriodStart2 = fertileWindowEnd.plusDays(1);
        LocalDate safePeriodEnd2 = currentPeriodDate.plusDays(periodCycleLength - 1);

        LocalDate pmsStart = currentPeriodDate.plusDays(periodCycleLength - 10);
        LocalDate pmsEnd = currentPeriodDate.plusDays(periodCycleLength - 1);

        LocalDate periodEndDate = currentPeriodDate.plusDays(periodDurationDays - 1);

        System.out.println("\n=== User Name ===: " + name);
        System.out.println("Here's a breakdown of your Menstrual Cycle");
        System.out.println("Your Cycle length is: " + periodCycleLength + " days");

        if (periodCycleLength >= 21 && periodCycleLength <= 35) {
            System.out.println("Your Period Cycle is within the normal range (21-35 days).");
            System.out.println("Here is your next 5 predicted period dates for the next 5 months:");

            LocalDate nextPeriodDate = currentPeriodDate;
            for (int i = 1; i <= 5; i++) {
                System.out.println("Period " + i + ": " + nextPeriodDate);
                nextPeriodDate = nextPeriodDate.plusDays(periodCycleLength);
            }
        } else {
            System.out.println("Your Cycle is outside the normal range.");
        }

        System.out.println("Your Ovulation date is: " + ovulationDate);
        System.out.println("Your Fertility Window (full): " + fertileWindowStart + " to " + fertileWindowEnd);
        System.out.println("Peak Fertility Days (most fertile): " + peakFertilePeriodStart + " to " + peakFertilePeriodEnd);
        System.out.println("Your Safe period before the start of your Fertile Window is: " + safePeriodStart1 + " to " + safePeriodEnd1);
        System.out.println("Your Safe period after the end of your Fertile Window is: " + safePeriodStart2 + " to " + safePeriodEnd2);
        System.out.println("PMS days are likely to begin between: " + pmsStart + " to " + pmsEnd);
        System.out.println("Your period will likely last: " + currentPeriodDate + " to " + periodEndDate);
        System.out.println("Your Estimated period duration is: " + periodDurationDays + " days");
    }

    public static void addUserDetails() {
        String name;

        while (true) {
            System.out.print("Please Enter your Name: ");
            name = input.nextLine().trim();

            boolean duplicateName = false;
            for (MenstruationCycle user : cycleDetails) {
                if (user.name.equalsIgnoreCase(name)) {
                    System.out.println("Error: User with the name '" + name + "' already exists. Please enter a different name.");
                    duplicateName = true;
                    break;
                }
            }

            if (!duplicateName) break;
        }

        LocalDate previousPeriodDate = null;
        LocalDate currentPeriodDate = null;
        int periodDurationDays = 0;

        while (true) {
            try {
                System.out.print("Enter your previous period date (YYYY-MM-DD): ");
                previousPeriodDate = LocalDate.parse(input.nextLine());
                break;
            } catch (DateTimeParseException e) {
                System.out.println("Invalid date format. Please use YYYY-MM-DD.");
            }
        }

        while (true) {
            try {
                System.out.print("Enter your current period date (YYYY-MM-DD): ");
                currentPeriodDate = LocalDate.parse(input.nextLine());

                if (currentPeriodDate.isBefore(previousPeriodDate)) {
                    System.out.println("Error: Current date must be after the previous date.");
                    continue;
                }
                break;
            } catch (DateTimeParseException e) {
                System.out.println("Invalid date format. Please use YYYY-MM-DD.");
            }
        }

        while (true) {
            try {
                System.out.print("Enter your typical period duration (in days): ");
                periodDurationDays = Integer.parseInt(input.nextLine());
                break;
            } catch (NumberFormatException e) {
                System.out.println("Invalid input for period duration. Please enter a valid number.");
            }
        }

        cycleDetails.add(new MenstruationCycle(name, previousPeriodDate, currentPeriodDate, periodDurationDays));
        System.out.println("Your details are currently being analyzed!! Please check on the View Menstruation Summary Section to view the summary of your Cycle.");
    }
    
    public static void viewMenstrualSummary() {
        if (cycleDetails.isEmpty()) {
            System.out.println("There are no Menstrual Records available.");
            return;
        }
        
        for (int i = 0; i < cycleDetails.size(); i++) {
            System.out.println("\nUser " + (i + 1));
            cycleDetails.get(i).showMenstrualCycle();
        }
    }
        
    public static void updateMenstruationCycle() {
        if (cycleDetails.isEmpty()) {
            System.out.println("There are no user records to update.");
            return;
        }
        
        System.out.print("Please Enter the user name to update Menstrual Summary: ");
        String indexName = input.nextLine().trim();
        
        MenstruationCycle foundUser = null;
        for(MenstruationCycle user : cycleDetails) {
            if(user.getName().equalsIgnoreCase(indexName)) {
                foundUser = user;
                break;
            }
        }
        
        if (foundUser == null) {
            System.out.println("User not found");
            return;
        }
        
        while (true) {
            try {
                System.out.print("Enter the new previous period date (YYYY-MM-DD): ");
                foundUser.previousPeriodDate = LocalDate.parse(input.nextLine());
                break;
            } catch (DateTimeParseException e) {
                System.out.println("Invalid date format. Please use YYYY-MM-DD.");
            }
        }
        
        while (true) {
            try {
                System.out.print("Enter the new current period date (YYYY-MM-DD): ");
                LocalDate newCurrentDate = LocalDate.parse(input.nextLine());
                if (newCurrentDate.isBefore(foundUser.previousPeriodDate)) {
                    System.out.println("Error: Current date must be after the previous date.");
                    continue;
                }
                foundUser.currentPeriodDate = newCurrentDate;
                break;
            } catch (DateTimeParseException e) {
                System.out.println("Invalid date format. Please use YYYY-MM-DD.");
            }
        }
        
        while (true) {
            try {
                System.out.print("Enter the new typical period duration (in days): ");
                foundUser.periodDurationDays = Integer.parseInt(input.nextLine());
                break;
            } catch (NumberFormatException e) {
                System.out.println("Invalid input for period duration. Please enter a valid number.");
            }
        }
        
        System.out.println("The User Menstrual Summary has been updated!!");
    }
}
